toolchain/install/busybox_info_test.cpp

```diff
@@ -166,5 +202,104 @@ TEST_F(BusyboxInfoTest, NotBusyboxSymlink) {
   EXPECT_FALSE(info.ok());
 }
 
+TEST_F(BusyboxInfoTest, LayerSymlinksInstallTree) {
+  auto actual_busybox = MakeFile(dir_ / "actual-busybox");
+
+  // Create a facsimile of the install prefix with even the busybox as a
+  // symlink. Also include potential relative sibling symlinks like `clang++` to
+  // `clang`.
+  auto prefix = MakeInstallTree(dir_ / "test_prefix", actual_busybox);
+
+  auto info = GetBusyboxInfo((prefix / "bin/carbon").string());
+  ASSERT_TRUE(info.ok()) << info.error();
+  EXPECT_THAT(info->bin_path, Eq(prefix / "bin/../lib/carbon/carbon-busybox"));
+  EXPECT_THAT(info->mode, Eq(std::nullopt));
+
+  info = GetBusyboxInfo((prefix / "lib/carbon/llvm/bin/clang").string());
+  ASSERT_TRUE(info.ok()) << info.error();
+  EXPECT_THAT(info->bin_path,
+              Eq(prefix / "lib/carbon/llvm/bin/../../carbon-busybox"));
+  EXPECT_THAT(info->mode, Eq("clang"));
+
+  info = GetBusyboxInfo((prefix / "lib/carbon/llvm/bin/clang++").string());
+  ASSERT_TRUE(info.ok()) << info.error();
+  EXPECT_THAT(info->bin_path,
+              Eq(prefix / "lib/carbon/llvm/bin/../../carbon-busybox"));
+  EXPECT_THAT(info->mode, Eq("clang++"));
+}
+
+TEST_F(BusyboxInfoTest, StopSearchAtFirstSymlinkWithRelativeBusybox) {
+  // Some install of Carbon under `opt`.
+  auto opt_prefix = MakeInstallTree(dir_ / "opt");
+
+  // A second install, but with its symlinks pointing into the `opt` tree rather
+  // than at its busybox.
+  MakeDir(dir_ / "lib/carbon");
+  MakeFile(dir_ / "lib/carbon/carbon-busybox");
+  MakeDir(dir_ / "bin");
+  auto target = MakeSymlink(dir_ / "bin/carbon", "../opt/bin/carbon");
+  MakeDir(dir_ / "lib/carbon/llvm/bin");
+  auto clang_target = MakeSymlink(dir_ / "lib/carbon/llvm/bin/clang",
+                                  opt_prefix / "lib/carbon/llvm/bin/clang");
+
+  // Starting from the second install uses the relative busybox rather than
+  // traversing the symlink further.
```

Comment:
Why is this important? If they're all symlinks to the same file, won't the user get the same behavior either way?

Reply:
The two installs might be different versions or unrelated.

Reply:
I don't follow how that could be the case, given that the test seems to say that one of them is a collection of symlinks to the other.

Reply:
We have two installs, let's say version v0.1 in `/usr` and v0.2 in `/usr/local`.

For some strange reason, there is also a symlink `/usr/local/carbon` -> `/usr/local/bin/carbon`.

There is no symlink from `/usr/local/...` into `/usr` (or vice versa).

The problem is that the symlink `/usr/local/carbon` (which points to `/usr/local/bin/carbon` and eventually to `/usr/local/lib/carbon/carbon-busybox`) could use the same path-relative traversal as `/usr/bin/carbon` and accidentally find `/usr/lib/carbon/carbon-busybox`. The test is checking that it does not do this.

While we don't really intend symlinks to exist in this way, the goal is to avoid confusing behavior if it does occur.

Reply:
It sounds like you're describing the behavior of `RejectSymlinkInUnrelatedInstall`, not `StopSearchAtFirstSymlinkWithRelativeBusybox`?

Reply:
I see, so the point is that even though in this case further traversal will still yield a correctly-shaped install tree, the non-traversal is important because of other cases where that's not true? If so, I think it'd be helpful to explain that in a comment here.

(It would be even better for the symlinks to point to something that's not a correctly-shaped install, so the test code directly illustrates the kind of case we're concerned about, but I don't think that's blocking.)